### DISCUSSION TITLE
Offline mark calculation

### DISCUSSION THREAD

**Discussion by arjunsuresh** [GATEOverflow/GATE-Marks-Evaluator#4] [Ideas]
Currently marks are calculates on fly using JS. We need to do offline mark calculation by parsing the downloaded response URL. For this a parser should be made which parses the the downloaded URL and returns the following data in a JSON format.

Question short ID, Question long ID, Question type, Question Mark, Response given, Actual answer, Subject ID, Obtained Mark

Here, 

Question short ID is the final number corresponding to each question image downloaded from the official URL and this is the same used in GO. This will be from c1 to c55 for technical and g1 to g10 for general aptitude.

Question long ID: The 9 digit question ID as given on the downloaded response for each question

Question type: NAT, MSQ or MCQ as given on response

Question Mark: 1 or 2 - c1 to c25 are 1 mark and c26 to c55 are 2 marks. g1 to g5 are 1 mark and g6 to g10 are 2 marks

Response given: Given answer as in the downloaded response

Actual answer and Subject ID to be taken from a Google Response sheet like this : 

- https://docs.google.com/spreadsheets/d/e/2PACX-1vRcxWRENCxZJe6aPTC2ospREhxroCzayF1KSTdSNyVVHmybITJ3CCzgcdZLPu2SOlSy8aw8KcpssgOi/pub?gid=0&single=true&output=csv

Obtained mark: This should return the positive mark if the response is evaluated correct as per the key and negative or 0 otherwise based on the question type. This should match the obtained mark shown in the Pragy's app.

**Comment by shbhmsingh72** [comment]
@arjunsuresh Is it possible to have a sample downloaded response sheet?

**Reply by arjunsuresh** [reply]
I dont have one to share at moment but if anyone allows sharing I'll do it.

**Comment by PritishC** [comment]
I am guessing that you plan to use this on a server so that you can get a faster response? Because the current JS version takes a while?

**Reply by arjunsuresh** [reply]
Never. Then site will crash 😛 JS does the calculation on the user side and that's the best here.

But this will be used offline for re-evaluation of marks like when say official key comes or for debated questions one can have a different view of results with a different key. The calculation of marks will be changed offline at some time using all the previous responses and then results updated.

**Comment by PritishC** [comment]
Is it possible to provide the answer key google sheet as a CSV file either via direct link or as a file on the filesystem? Downloading the answer key from Google Sheets is a hassle that involves Google Drive API and can't be done directly.
For the response sheet, I am assuming it will be available using the TCS digialm direct link.
PS: Please let us download the answer key google sheet for testing.

**Reply by arjunsuresh** [reply]
I have made the sheet link a CSV now. And copy paste is enabled on the page - so you can do whatever you want using the data :) 
But there is already a Pull request here for this and it looks good.

**Reply by PritishC** [reply]
Well, it's a competition as you said, so you get to choose the best. I hope you won't discount the rest of us based on one submission.

I am almost done, but came up with some trouble. When we extract images to find the question short ID, it appears that my response sheet goes uptil Q58 in the CS section, but the last should be Q55, which is bizarre. I'll send a PR.

**Reply by arjunsuresh** [reply]
Yes, that I too have no idea why :) Instead of 54 it is 58 for one set.

**Comment by manishbhatt151** [comment]
Hi @arjunsuresh , I have created a PR for the parser [https://github.com/GATEOverflow/GATE-Marks-Evaluator/pull/5]. Please review and let me know if there are any issues or questions!
